lib/websocket/connection.js

```diff
@@ -0,0 +1,413 @@
+'use strict'
+
+// TODO: crypto isn't available in all environments
+const { randomBytes, createHash } = require('crypto')
+const { Blob } = require('buffer')
+const { uid, states, opcodes } = require('./constants')
+const {
+  kReadyState,
+  kResponse,
+  kExtensions,
+  kProtocol,
+  kBinaryType,
+  kWebSocketURL,
+  kController
+} = require('./symbols')
+const { fireEvent } = require('./util')
+const { MessageEvent } = require('./events')
+const { makeRequest } = require('../fetch/request')
+const { fetching } = require('../fetch/index')
+const { getGlobalDispatcher } = require('../..')
+
+/**
+ * @see https://websockets.spec.whatwg.org/#concept-websocket-establish
+ * @param {URL} url
+ * @param {string|string[]} protocols
+ * @param {import('./websocket').WebSocket} ws
+ */
+function establishWebSocketConnection (url, protocols, ws) {
+  // 1. Let requestURL be a copy of url, with its scheme set to "http", if url’s
+  //    scheme is "ws", and to "https" otherwise.
+  const requestURL = url
+
+  requestURL.protocol = url.protocol === 'ws:' ? 'http:' : 'https:'
+
+  // 2. Let request be a new request, whose URL is requestURL, client is client,
+  //    service-workers mode is "none", referrer is "no-referrer", mode is
+  //    "websocket", credentials mode is "include", cache mode is "no-store" ,
+  //    and redirect mode is "error".
+  const request = makeRequest({
+    urlList: [requestURL],
+    serviceWorkers: 'none',
+    referrer: 'no-referrer',
+    mode: 'websocket',
+    credentials: 'include',
+    cache: 'no-store',
+    redirect: 'error'
+  })
+
+  // 3. Append (`Upgrade`, `websocket`) to request’s header list.
+  // TODO: what if the headersList already has this header?
+  // in the browser, this header is filtered, but here it isn't.
+  request.headersList.append('upgrade', 'websocket')
```

Comment:
Hey, do we need this? currently we are sending 2 upgrade headers:

<img width="647" alt="Screenshot 2022-12-04 at 15 13 44" src="https://user-images.githubusercontent.com/14154314/205495545-6acdc8a5-1baf-4d38-b9bf-9fadebc7637c.png">
de headers:

Reply:
Weird, I'm unsure why this happens (the `upgrade` header was previously a forbidden header). Maybe setting `upgrade: websocket` in the dispatcher option sets this header?

Reply:
Yes it does. https://github.com/nodejs/undici/blob/68c269c4144c446f3f1220951338daef4a6b5ec4/lib/client.js#L1397